### DISCUSSION TITLE
Share to: HTTP Post + Media File

### DISCUSSION THREAD

**Discussion by JpEncausse** [Waboodoo/HTTP-Shortcuts#394] [Q&A]
Thanks for this amazing tool !
I use Shortcuts to "share to" an AirTable Database like a WebClipper.
From any app I get the "Shared stuff" + a prompt for Title and Description (filled from clipboard).
But I found few limitation or don't know how to handle it ?

- How do you set Variable order ? It ask Description before Title
- Is it possible to grab something on the screen (like WebClipper would do) I don't think so
- Is it possible to prompt for Text with a cancel that would stop the flow (because I forgot to Select/Copy the text)

The more interesting part: Many app with photo or audio (like the recorder) share the file itself ! But the Shortcut app display an erreur close to : "There is no link with share to for this content"

Would it be possible to share to file ? 
As for now, I use POST request with JSON content but maybe I can use form upload ? but in the options there is no "files from the clipboard".

Any ideas to build that kind of universal app clipper would be very welcome !

**Comment by Waboodoo** [comment]
> How do you set Variable order ? It ask Description before Title

Currently, there's no simple way to define the order in which variables are resolved. I will change this for the next version so that they are resolved in the same order as they appear on the Variables screen, such that you can change the order by re-arranging them there.

> Is it possible to grab something on the screen

No, this is not possible

> Is it possible to prompt for Text with a cancel that would stop the flow

You could achieve this via the Scripting feature. Assuming you have a variable "myVariable" and you want to abort the execution if it doesn't have any text, you could do this by opening the shortcut editor, going to the Scripting section and in the "Run before execution" field enter code like this:

```js
if (getVariable("myVariable") === "") {
    abort();
}
```

or if you wanted to prompt for text in this case, you could do that with code like this instead:

```js
if (getVariable("myVariable") === "") {
    const text = prompt("Please enter text");
    if (!text) {
        abort();
    }
    setVariable("myVariable", text);
}
```

More on Scripting here: https://http-shortcuts.rmy.ch/scripting

> Would it be possible to share to file ?

Yes, this is possible, but with some limitations. If you use a form upload instead of a JSON body for your request, you can add file parameters. This will allow you to share files into the app and use them as part of the request. This may however not be compatible with also sharing text at the same time. There is no option to take a file from the clipboard. More on file sharing here: https://http-shortcuts.rmy.ch/advanced#share-files

**Comment by JpEncausse** [comment]
Thanks, not sure to understand the last part :
Let say I'm in the Android Recorder, I click "Share to" (the file) and select Shortcut.

In the Shortcut I need
- to configure a Sharable variable
- to setup a form upload

In my case it seems 
- the shareable variable only accept text
- and the Share Files form upload only accept files from filesystem or predefined

Is it possible and how to use the file shared by "share to" feature in the form upload. I missed something or is it the part that is not possible ?

**Reply by Waboodoo** [reply]
There are 2 types of sharing that the app supports: text and files.
- When sharing text, any variable for which you have enabled sharing can receive the shared value. It will use that value instead of its normal value. If you have multiple shortcuts that use such variables, you'll be asked to select which shortcut should be run.
- When sharing files, any shortcut which uses a file body or file parameter can receive the shared file. It will use that file instead of opening the file picker or using the pre-configured file. If you have multiple shortcuts that use such file body or file parameters you'll be aksed to select which shortcut should be run.

There is no way to share both text and files, so if you use both a share-enabled variable and a file body or parameter in a shortcut, only one or the other but not both can make use of the sharing feature, and the other one will use its normal behavior (i.e., as if you had manually started the shortcut).

If the file that you are sharing happens to be a text file, it will use the text-sharing behavior, not the file-sharing behavior, maybe that's the case here? There is currently no way to change this, but I could look into it if this turns out to be a problem.

**Comment by JpEncausse** [comment]
Sorry I might be stupid, I'm missing somthing:

> When sharing files, any shortcut which uses a file body or file parameter can receive the shared file. It will use that file instead of opening the file picker or using the pre-configured file.

Ok it works, 
- variable with sharing, without JSON encoding 
- adding header content-type with application/octet-stream
- filechooser upload option
- add the text variable in the URL

It works with a Node-RED server but not with Airtable. I will route it correctly :-)
With images (photo) shared from GooglePhoto my server answer payload too large
